pkg/suggestion/NAS_Reinforcement_Learning/Operation.py

```diff
@@ -0,0 +1,81 @@
+import itertools
+import numpy as np
+from pkg.api.python import api_pb2
+
+
+class Operation(object):
+    def __init__(self, opt_id, opt_type, opt_params):
+        self.opt_id = opt_id
+        self.opt_type = opt_type
+        self.opt_params = opt_params
+    
+    def get_dict(self):
+        opt_dict = dict()
+        opt_dict['opt_id'] = self.opt_id
+        opt_dict['opt_type'] = self.opt_type
+        opt_dict['opt_params'] = self.opt_params
+        return opt_dict
+
+    def print_op(self):
+        print("Operation ID: \n\t", self.opt_id)
+        print("Operation Type: \n\t", self.opt_type)
+        print("Operations Parameters:")
+        for ikey in self.opt_params:
+            print("\t {}: {}".format(ikey, self.opt_params[ikey]))
+
+
+class SearchSpace(object):
+    def __init__(self, operations):
+        self.operation_list = list(operations.operation)
+        self.search_space = list()
+        self._parse_operations()
+        print()
+        self.num_operations = len(self.search_space)
+    
+    def _parse_operations(self):
+        # search_sapce is a list of Operation class
+
+        operation_id = 0
+
+        for operation_dict in self.operation_list:
+            opt_type = operation_dict.operationType
+            opt_spec = list(operation_dict.parameter_configs.configs)
+            # avail_space is dict with the format {"spec_nam": [spec feasible values]}
+            avail_space = dict()
+            avail_space_len = list()
+            num_spec = len(opt_spec)
+
+            for ispec in opt_spec:
+                spec_name = ispec.name
+                if ispec.parameter_type == api_pb2.CATEGORICAL:
+                    avail_space[spec_name] = list(ispec.feasible.list)
+                elif ispec.parameter_type == api_pb2.INT:
+                    spec_min = int(ispec.feasible.min)
+                    spec_max = int(ispec.feasible.max)
+                    spec_step = int(ispec.feasible.step)
+                    avail_space[spec_name] = range(spec_min, spec_max+1, spec_step)
+                elif ispec.parameter_type == api_pb2.DOUBLE:
+                    spec_min = float(ispec.feasible.min)
+                    spec_max = float(ispec.feasible.max)
+                    spec_step = float(ispec.feasible.step)
+                    if spec_step == 0:
+                        print("Error, NAS Reinforcement Learning algorithm cannot accept continuous search space!")
+                        exit(999)
```

Comment:
Yes. If the definition of a study is invalid, there is no way we could start a task.

Reply:
so If I submit a invalid studyjob, suggestion service exit; Then I submit another valid studyjob, but nothing can work since I cannot get suggestion from (suggestion service).
I think raising a exception makes sense for a invalid studyjob

Reply:
I think we should check that StudyJob is valid in initialize part in the StudyJob controller. Just like you notice here: https://github.com/kubeflow/katib/issues/372. 
What do you think about extending ```validateStudy``` function here: https://github.com/kubeflow/katib/blob/f11c13e8c65a9236afe2878e5c915a1c769ad52b/pkg/controller/studyjob/katib_api_util.go#L28 ?

Reply:
agree. so suggestion service can always take each studyjob as valid one.

Reply:
Ok. I will make this changes tomorrow.

Reply:
I see. I will remove the checking here after the validateStudy is modified and merged.